From d5f783be1aac00812050cbc1af7c43f857652598 Mon Sep 17 00:00:00 2001
From: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
Date: Thu, 6 Feb 2014 04:53:53 +0000
Subject: [PATCH 109/351] extcon: arizona: Give option to declare an mic
 open-circuit as a 4-pole

Normally the driver will declare headphones for an open-circuit, whilst
this is still the most sensible choice. Provide provision to declare
open-circuit as a microphone for those that wish.

Change-Id: Ic7dca9581cd94fb7204a37aa37f06cd5d4b470fc
Signed-off-by: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
Signed-off-by: Nikesh Oswal <Nikesh.Oswal@wolfsonmicro.com>
---
 drivers/extcon/extcon-arizona.c   |    6 +++++-
 include/linux/mfd/arizona/pdata.h |    3 +++
 2 files changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/extcon/extcon-arizona.c b/drivers/extcon/extcon-arizona.c
index da925b8..4031be8 100644
--- a/drivers/extcon/extcon-arizona.c
+++ b/drivers/extcon/extcon-arizona.c
@@ -878,7 +878,7 @@ static void arizona_micd_detect(struct work_struct *work)
 	/* Due to jack detect this should never happen */
 	if (!(val & ARIZONA_MICD_STS)) {
 		dev_warn(arizona->dev, "Detected open circuit\n");
-		info->mic = false;
+		info->mic = arizona->pdata.micd_open_circuit_declare;
 		arizona_stop_mic(info);
 		arizona_identify_headphone(info);
 		info->detecting = false;
@@ -1208,6 +1208,10 @@ static int arizona_extcon_get_pdata(struct arizona *arizona)
 			of_property_read_bool(arizona->dev->of_node,
 					      "wlf,micd-software-compare");
 
+	pdata->micd_open_circuit_declare =
+		of_property_read_bool(arizona->dev->of_node,
+				      "wlf,micd-open-circuit-declare");
+
 	pdata->jd_gpio5 = of_property_read_bool(arizona->dev->of_node,
 						"wlf,use-jd-gpio");
 
diff --git a/include/linux/mfd/arizona/pdata.h b/include/linux/mfd/arizona/pdata.h
index 7aa4c1e..c8d0386 100644
--- a/include/linux/mfd/arizona/pdata.h
+++ b/include/linux/mfd/arizona/pdata.h
@@ -163,6 +163,9 @@ struct arizona_pdata {
 	/** Force MICBIAS on for mic detect */
 	bool micd_force_micbias;
 
+	/** Declare an open circuit as a 4 pole jack */
+	bool micd_open_circuit_declare;
+
 	/** Delay between jack detection and MICBIAS ramp */
 	int init_mic_delay;
 
-- 
1.7.9.5

